stdlib/public/Platform/musl.modulemap

```diff
@@ -2,110 +2,876 @@
 //
 // This source file is part of the Swift.org open source project
 //
-// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
 // Licensed under Apache License v2.0 with Runtime Library Exception
 //
 // See https://swift.org/LICENSE.txt for license information
 // See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 //
 //===----------------------------------------------------------------------===//
 
-module SwiftMusl [system] {
-  header "stdc-predef.h"
+module _musl [system] {
+  config_macros _GNU_SOURCE, _BSD_SOURCE, _XOPEN_SOURCE, _POSIX_SOURCE, _POSIX_C_SOURCE, _LARGEFILE64_SOURCE, _ALL_SOURCE, _DEFAULT_SOURCE, __STRICT_ANSI__
+
   header "features.h"
 
-  // C standard library
+  header "bits/musldefs.h"
+
+  exclude header "bits/alltypes.h"
+
+  // Type declarations live in bits
+  module types {
+    umbrella "bits/types"
+      module * {
+      export *
+    }
+  }
+}
+
+module stdc_assert [system] {
```

Comment:
If we do it this way, then we can drop all of the platform checks and just `import stdlib_h` etc from Swift once we get the rest of the module maps converted.

Reply:
Interesting. I don't think we want the Swift interface to the C library to be things like `import stdlib_h` though. That seems like the wrong approach. `import Libc` would be my personal preference, and maybe `import Posix` and `import LibcExt` also, the idea being that `Libc` will get you a standard set of Libc declarations (ones from the C standard), `Posix` will get you a standard set of Posix declarations (possibly even on Windows we could provide some subset there), and `LibcExt` would get you any platform extensions.

I have some ideas about how to get there; it's a little more complicated than just dropping module maps in the right places, obviously, because the entire point is that the _Swift_ interface for `Libc` and `Posix` would be standardised (e.g. nullability annotations on the C declarations would be the same, regardless of what the underlying C library has done).

Reply:
As regards the first issue, that's presumably a side effect of not having `-fbuiltin-headers-in-system-modules`? With that flag, it'll find the correct header, but without it we're going to need your workaround.

Reply:
No `-fbuiltin-headers-in-system-modules` isn't involved. `header` declarations always resolve relative to the module map file and do not use header search paths. Otherwise the OS/SDK module would accidentally cover the libc++ or clang stdint.h instead of its own stdint.h. module maps can't be expected to know about other headers with the same name in different directories, or how those layer with everything else.

I would still underscore the modules for the real headers so that those don't show up in autocomplete.

Personally I prefer having cover modules for each of the headers because they aren't that related to each other. e.g. I might need something out of string.h or time.h but I don't want to pull in anything from signal.h or stdlib.h. There's also the performance concern: a `Libc` module will end up containing a lot of headers, e.g. on Apple platforms just stdlib.h pulls in 75 headers. That's a lot of headers to stat and open when all someone wanted was e.g. `size_t`, and it gets real expensive when NFS gets involved. POSIX is much worse, and probably shouldn't be used all that often from ObjC or Swift anyway.

That's my opinion at least. I get the convenience of having to import only one or two things and you get everything you need, but you also get a huge pile of junk in your autocomplete that you don't need, and it slows down builds too.

Reply:
> If we do it this way, then we can drop all of the platform checks and just `import stdlib_h` etc from Swift once we get the rest of the module maps converted.

We'd have to fix a few more things before that would work, particularly relating to the Clang Importer (it would need to guarantee to import the C library in a uniform manner in spite of C-source-compatible differences in declarations on different platforms).

Reply:
I've adopted the same approach (the `_modules/<modulename>.h` one, I mean) in the musl modulemap, as that seems like a neat idea.